File: measurements/engines-keep-changing/family.py
# The test card through the Python engine: pip install spintax-core==0.5.0 (the version this
# article's table reports; drop the pin to score whatever is current), then python family.py
# Prints one line per case: its id and the share of renders that come out right (as measure.mjs scores it).
import json
import re
from importlib.metadata import version
from pathlib import Path

from spintax_core import render

SEEDS = 100


def distinct(out: str) -> bool:
    """A permutation shuffles distinct elements, which no regex alternation can say."""
    items = re.split(r",\s*|\s+and\s+", re.sub(r"^[^:]*:\s*", "", out).rstrip("."))
    return len(set(items)) == len(items)


card = json.loads((Path(__file__).parent / "card.json").read_text(encoding="utf-8"))["cases"]
print("spintax-core", version("spintax-core"))
for c in card:
    ctx = c.get("context", {})
    if "expect" in c:
        out = render(c["src"], context=ctx, seed=1)
        score = 1.0 if out == c["expect"] else 0.0
    else:
        shape = re.compile(c["shape"])
        outs = [render(c["src"], context=ctx, seed=s) for s in range(1, SEEDS + 1)]
        out = outs[0]
        score = sum(1 for o in outs if shape.fullmatch(o) and distinct(o)) / SEEDS
    print(f"{c['id']:<26}{score:.2f}  {json.dumps(out, ensure_ascii=False)}")

File: measurements/engines-keep-changing/family.php
<?php
// The test card through the PHP engine: composer require spintax/core:0.10.0 (the version this
// article's table reports; drop the pin to score whatever is current), then php family.php
// Prints one line per case: its id and the share of renders that come out right (as measure.mjs scores it).
// The PHP engine takes no seed — the random picks are seeded here with mt_srand, which is why a
// shape (not an exact string) is what a random case is scored against, as in measure.mjs.
require __DIR__ . '/vendor/autoload.php';

use Spintax\Core\Render\Pipeline;

const SEEDS = 100;

// A permutation shuffles distinct elements, which no regex alternation can say.
function distinct(string $out): bool
{
    $items = preg_split('/,\s*|\s+and\s+/u', rtrim(preg_replace('/^[^:]*:\s*/u', '', $out), '.'));
    return count(array_unique($items)) === count($items);
}
$card = json_decode(file_get_contents(__DIR__ . '/card.json'), true)['cases'];
foreach (json_decode(file_get_contents(__DIR__ . '/vendor/composer/installed.json'), true)['packages'] as $p) {
    if ($p['name'] === 'spintax/core') echo "spintax/core {$p['version']}\n";
}
$pipeline = new Pipeline();
foreach ($card as $c) {
    $ctx = $c['context'] ?? [];
    $render = function (int $seed) use ($pipeline, $c, $ctx): string {
        mt_srand($seed);
        return $pipeline->render($c['src'], $ctx);
    };
    if (array_key_exists('expect', $c)) {
        $score = $render(1) === $c['expect'] ? 1.0 : 0.0;
    } else {
        $shape = '/' . str_replace('/', '\/', $c['shape']) . '/u';
        $ok = 0;
        for ($seed = 1; $seed <= SEEDS; $seed++) {
            $out = $render($seed);
            $ok += (preg_match($shape, $out) && distinct($out)) ? 1 : 0;
        }
        $score = $ok / SEEDS;
    }
    printf("%-26s%.2f  %s\n", $c['id'], $score, json_encode($render(1), JSON_UNESCAPED_UNICODE));
}
